### ISSUE TITLE
Support for maps of more complex types in set_union

### ISSUE THREAD

**Issue by bwspenc** [libMesh/TIMPI#69]
I'd like to be able to pass a map of more complex data types to `communicator::set_union()`. Specifically, I'm interested in passing in the following type from a MOOSE code:

`std::map<unsigned int, std::vector<RealVectorValue>>`

It looks like currently there is no support for `std::map<unsigned int, std::vector<anything>>`.

**Comment by bwspenc** [comment]
@austinltam

**Comment by loganharbour** [comment]
```
  std::map<unsigned int, std::vector<RealVectorValue>> some_map;
  comm().set_union(some_map);
```

Compiles for me no problem. Did you include `parallel_sync.h` ?

**Comment by bwspenc** [comment]
I don't have ready access to the specific code, but @austinltam will look into whether that header include was missing this evening. We actually spent a while looking into this with @roystgnr. Hopefully it's as easy as including another header!

**Comment by austinltam** [comment]
I included parallel_sync.h but I am still getting the buffer_type error as shown below.

 error: no type named 'buffer_type' in 'struct libMesh::Parallel::PairBufferTypeHelper<unsigned int, false, std::vector<libMesh::VectorValue<double>, std::allocator<libMesh::VectorValue<double> > >, false>'
  138 |                                         Has_buffer_type<Packing<T2>>::value>::buffer_type buffer_type;

**Comment by roystgnr** [comment]
Try adding the macro invocation in https://github.com/libMesh/TIMPI/pull/70/commits/3141b802a26b07afcadff2db3d44f2d9a14ca117

and see if that works for you?

**Comment by roystgnr** [comment]
Ping?  I finally got around to fixing the name collision in https://github.com/libMesh/TIMPI/pull/70 and I'd like to start getting it moved into libMesh and Moose if it works for you.  I have vacation coming up soon and I'm hoping to get everything I've been working on this month pushed downstream first, with enough advance notice to fix any problems that occur.

**Comment by bwspenc** [comment]
@roystgnr Sorry it took me so long to get to this!  I just had a chance to test the code in #70, and verified that it works in our MOOSE code. I tested it by doing the following:

1) I added the following code to the beginning of the XFEMMaterialStateMarkerBase::finalize() method (in moose/modules/xfem/src/userobjects), which is where the code is that we want this feature for:
```
  unsigned int pid = processor_id();
  std::cout<<"BWS0 pid: "<<pid<<std::endl;
  RealVectorValue rvv;
  rvv(0) = rvv(1) = rvv(2) = pid;
  std::map<unsigned int, std::vector<RealVectorValue>> tmp;
  tmp[pid] = {rvv};
  _communicator.set_union(tmp);
  for (auto t: tmp)
    std::cout<<"BWS1 pid: "<<t.second[0](0)<<std::endl;
```

2) I ran the XFEM module executable in parallel with this change on a problem in modules/xfem/test/tests/solid_mechanics_basic that uses this class, and captured the output in separate files per processor:
```
mpiexec -n 2 ~/projects/moose/modules/xfem/xfem-opt -i crack_propagation_2d.i --keep-cout --redirect-stdout
```

3) When I examine the output from the output streams, I see this from processor 0:
```
BWS0 pid: 0
BWS1 pid: 0
BWS1 pid: 1
```
and this from processor 1:
```
BWS0 pid: 1
BWS1 pid: 0
BWS1 pid: 1
```

This confirms that the `set_union()` call worked on a `std::map<unsigned int, std::vector<RealVectorValue>>`. I tried this without your fix, and it compiles, but I get a runtime error.
